File: app/src/main/java/ilapin/opengl_research/ObjectsPool.kt
package ilapin.opengl_research

import java.util.*

/**
 * @author raynor on 27.01.20.
 */
class ObjectsPool<T>(private val createObject: () -> T) {

    private val pool = LinkedList<T>()

    fun obtain(): T {
        return if (pool.isEmpty()) {
            createObject()
        } else {
            pool.removeFirst()
        }
    }

    fun recycle(o: T) {
        pool += o
    }
}

File: app/src/main/assets/scenes/objects_pool.js
function ObjectsPool(createObject) {

    this.createObject = createObject;

    this.pool = Array();

    this.obtain = function() {
        if (this.pool.length == 0) {
            return this.createObject();
        } else {
            return this.pool.shift();
        }
    };

    this.recycle = function(object) {
        this.pool.push(object);
    };
}